### DISCUSSION TITLE
組合語言的一個小坑

### DISCUSSION THREAD

**Discussion by FakeWood** [1101-GS4719-C/Python-Community-GS4719-C#1878] [General]
其實也不算坑，就是系上老師上課沒講清楚而已

這學期修了組合語言，老師在上FPU的章節時，跟我們說FPU也有CPU的CMP這類的指令可以搭配JE、JB之類的使用。
但我在做期末Project的時候才發現，FPU做完比較後不會動到CPU的Flags，所以用conditional jump會有錯，
所以用FPU做完比較後，應該要先把FPU裡的Flags放到暫存器AX裡，再從AX裡移到CPU的Flags裡，才可以用conditional jump

舉我的例子來說
![image](https://github.com/NCU-GS4719-Python/Python-Community/assets/105701699/6cac5dc2-0665-4aff-be85-5b980484578f)
在我做完 `ftst` 後，要用 `fstsw ax` 把FPU裡的Flags放到AX裡，再用 `sahf` 把AH的東西丟給CPU的Flags，這樣才可以用 `jne`

以上是做組語專案的一點小心得

**Comment by TobywithDino** [comment]
我自己是在做組語專案的時候以為不能處理三角函數，而使用python輸出一個tangent值的查表，後來弄一弄發現，X !，組語的FPU中可以得到SIN跟COS值

**Comment by yubo0oo** [comment]
我覺得，這學期修組合語言的課，遇到這種情況真的是讓人頭疼。老師上課時沒有講清楚FPU和CPU的Flags之間的區別，讓我們在做期末Project時才發現問題。

在做FPU的比較後，發現它不會動到CPU的Flags，所以用conditional jump會出錯。這真的是一個很大的坑，但其實也是學習過程中的寶貴經驗。我在做完FPU的比較後，才知道要先把FPU裡的Flags放到暫存器AX裡，再從AX裡移到CPU的Flags裡，才能正確使用conditional jump。

舉我的例子來說，在做完 `ftst` 後，需要用 `fstsw ax` 把FPU的Flags放到AX裡，再用 `sahf` 把AH的東西丟給CPU的Flags，這樣才可以正確地用 `jne` 進行條件跳轉。

這次的經歷讓我深刻理解了FPU和CPU之間的運作差異，也學到了如何正確地處理這種情況。我覺得這些小心得對於以後的組語專案會非常有幫助。希望大家在遇到類似問題時，也能夠找到解決的辦法，共同進步！你們有沒有遇到過類似的情況呢？歡迎分享你們的經驗和心得！
